### ISSUE TITLE
Paths for Struct Representation Tuple

### ISSUE THREAD

**Issue by hannahhoward** [ipld/specs#220]
If I create a scheme where I define a struct type with a touple type, does that remove my ability to look up fields by name when constructing a path? IOW, if I have:

```ipldsch
type Thing struct {
    prop String
    otherProp String
} representation tuple
```

Is prop only referenced as `/thing/0` and otherProp only `/thing/1` -- If for example, I was building a selector?

This is just a question -- not sure if it needs clarification but I thought it would be good to have whatever the answer is get documented in comments here for posterity.

**Comment by rvagg** [comment]
I don't think we've clarified how paths work in conjunction with schemas yet, I'm sure @warpfork has more mature thoughts on this though. There are issues in a number of places where the schema will present a different lens on the data than the data model will. The most egregious of this will be advanced layouts where the underlying data may be entirely foreign (think encrypted containers!). But even unions and all of the non-default representations present challenges. My understanding is that we're going to have to build selector modes to operate at the different levels where appropriate. I'm assuming context will drive this but I suspect we're going to have to deal with some messiness and aim toward a sensible user-facing model.

**Comment by warpfork** [comment]
Yep, what @rvagg said -- if the data is handled via schema, it should more or less feel like a lens that just makes it possible to see the data with field names even though the representation might be as list.

In the terminology of the [go library API](https://godoc.org/github.com/ipld/go-ipld-prime/), if we parse some data without a schema, we'd get an `ipld.Node` that answers `.Kind()` as "list"...

and if we parse that same data *with* that schema (or, equivalently, wrap the already parsed nodes in a wrapper interface that does the same thing, but as a decorator instead of via re-parse), we'll get an `ipld.Node` that answers `.Kind()` as "map".  Correspondingly, all the `LookupByString`, etc, methods which work on maps should... well, work. :)

In the latter case, the `ipld.Node` will *also* implement [`typed.Node`](https://github.com/ipld/go-ipld-prime/blob/8dff1cbec43bfe13de81a1e4c85b770d8bd4f4c5/impl/typed/typedNode.go#L8-L58).  So, we could ask `.Type().Kind()` on it, and *that* will say "struct".  And `.Representation()` will return another `ipld.Node` which corresponds to the representation and has the appropriate behaviors -- so `.Representation().Kind()` will say "list", and that node will be traversable by index, and so on.

**Comment by warpfork** [comment]
So, to focus back to the original question: it follows that **yes**, pathing should work by field when you parse data like that with a schema like that.

And the libraries should do most of this automatically.  E.g., the [traversal](https://godoc.org/github.com/ipld/go-ipld-prime/traversal) and [selectors](https://godoc.org/github.com/ipld/go-ipld-prime/traversal/selector) libraries don't actually even need a special understanding of schemas or types in order to do this -- they just naturally do the right thing because the schema'd `Node` interface already presents things to them such that it all transparently does the right thing.

**Comment by hannahhoward** [comment]
Ok! This is mostly interesting to me in terms of Graphsync, which is used in FileCoin, where people are heavily relying on cbor-gen (@whyrusleeping's library), which encodes everything as tuple. While there is a lot of code still to be written, it sounds like if we make typed nodes for relevant data (having implemented representation tuple) we can then use selectors that look up things by name rather than index, which is ergonomically quite difficult. That's a lot of ifs but it's nice to know the intent is to get there. There is ongoing discussions over whether tuple is our final representation vs map, but still good to know this is an option.
